### ISSUE TITLE
Uploader tool in XCA without client-side validation with authentication tokens

### ISSUE THREAD

**Issue by phraenquex** [m2ms/fragalysis-frontend#1226]
@tdudgeon / @kaliif must define scope.  It's the tail-end of the XCA uploader tool.

Dependency is #1196.

**Comment by mwinokan** [comment]
New tool to aid the speed of data dissemination that does:

- The loader validation offline from the stack (so that it's quicker)
- Upload of the tarball to the chosen stack

**Comment by mwinokan** [comment]
@kaliif says there is a lot of work (~1 week) to implement the validation on the client side but it is progressing.

Kalev is currently implementing the yaml validation in the XCA client. The client first requests the schema from the server and check whether the yaml's are formatted correctly and have all the correct attributes/keys.

Since the presence/integrity of the files will still need to be checked on the server post upload, where errors are thrown as the data are loaded/processed.

@phraenquex is less concerned about the size/difficulty of the first upload as there may always be issues to work out.

**@phraenquex raises concerns about duplicating the validation effort and it is suggested that @kaliif starts with implementing just the target uploading without having the client-side validation**

**Comment by kaliif** [comment]
@phraenquex, @mwinokan, @Waztom, @tdudgeon, an alternative to consider - I got the automated CAS login from the script working using a headless browser (Playwright). It's heavier than an average dependency, but it's installable from PyPI so it can be hidden from the user. When running the upload script, the user has to specify CAS credentials something like 
`--username=<cas_username> --password=<cas_password>` and off you go.
It requires some backend changes, Playwright doesn't yet support streaming uploads so it has to be done in chunks, something that the backend currently doesn't support. But it's not too big of a change and on the plus side, this will allow implementing resuming interrupted uploads.

**Comment by phraenquex** [comment]
@tdudgeon your suggestion of a token sounds sound.  One detail:  the B/E should provide (via the F/E) a _special_ token, which expires soon etc.  That then is the thing the user provides to the Uploader.

@mwinokan what do you think:  does this faff actually simplify anything for the user?

**Comment by mwinokan** [comment]
@phraenquex I think that a username and password for CAS will be simpler for the users, and @kaliif's suggestion which would allow for resuming of interrupted uploads sounds great

**Comment by phraenquex** [comment]
No, it must be generic and managed within Fragalysis stack, for anybody (in future) running their own stack, adapted to whatever authentication mechanism they use (i.e. not CAS).

**Comment by mwinokan** [comment]
@kaliif and @alanbchristie please use the authentication middleware to allow for short-term (1 hour) authentication tokens to be requested by authenticated users.

Then @matej-vavrek can add this to the hamburger menu

**Comment by mwinokan** [comment]
@kaliif has confirmed that an implementation with authentication tokens is tractable

**Comment by mwinokan** [comment]
@phraenquex asks how this will streamline the data release process, @mwinokan says this does help by removing the need to copy the tarballs locally (issues with bad internet connections), @tdudgeon also says that snapshotting raw data for the XCA tests will benefit from this (#1588).

@kaliif says it is no longer a daunting task and @phraenquex suggests proceeding

**Comment by mwinokan** [comment]
@kaliif is working on this and fixing some final glitches

**Comment by mwinokan** [comment]
@kaliif has finished development, and the `/api/token` endpoint on the staging stack is implemented.

@tdudgeon is unable to deploy it to the XCA staging environment while his FedID is not working.

To discuss at hackathon

**Comment by mwinokan** [comment]
Thanks @kaliif please could you add the following:

- CLI nicknames for the production and staging stack URLs. So the user can either provide `staging`, `production` or the URL to another stack
- Tarball creation at the start of the uploader command. The syntax for the name should be `<target_name>_v<data_version>_upload_<upload_number>_<date>.tgz`
- Add a CLI option to use an existing tarball (default is to create one from the latest compatible data)

**Comment by mwinokan** [comment]
@kaliif sorry forgot to ask you in the meeting, have the [above](https://github.com/m2ms/fragalysis-frontend/issues/1226#issuecomment-2609602612) changes been implemented?

**Comment by kaliif** [comment]
@mwinokan no, thanks for the reminder.

Do you have an example of how the data are laid out on the disk and how/which directory the user is likely to work in? The uploader needs to find the data somehow. Any relevant environment variables?

Adding nicknames is easy enough but it contradicts Frank's position on coupling the tools to concrete instances -- he was adamant that the uploader should not use the CAS login but must be generic for other deployments. Hardcoding staging and prod addresses violate that principle. Thoughts?

**Comment by mwinokan** [comment]
@kaliif please see the code in the collator/aligner for determining the correct working directory (or chat to @tdudgeon).

The relevant directory structure will be something like this

```
working_dir/upload-current/upload_1
```

Please create the tarball at:

```
working_dir/upload-current/<target_name>_v<data_version>_upload<upload_number>_<date>.tgz
```

> Hardcoding staging and prod addresses violate that principle.

I agree, and great point. Please could you ask @tdudgeon to maintain some sort of JSON/YAML file at the root of the DLS XCA deployment containing a mapping from nickname to URL for the stacks and use this if it can be found by the uploader tool.

**Comment by mwinokan** [comment]
@kaliif is working on the last changes, and should be wrapped up today. 

@tdudgeon's input is needed for specifying the URL nicknames, environment variables are suggested instead of JSON

**Comment by kaliif** [comment]
@mwinokan discussed this with Tim today and he pointed out that xchem-align creates an environment when set up, which if I understood correctly is different for staging and prod so effectively you only upload to single stack, prod or staging and not both. Is this correct? I understood previously that there should be different options always available and you'd specify them as `--url staging` or `--url production`.
If it is the former, `url` option can be silent, there's only a default address and you'd need to override it only if you really want to upload to a different stack.

**Comment by tdudgeon** [comment]
The two XCA environments are not specifically tied to any particular Fragalysis environment so I think the `--url staging` or `--url production` approach is correct. Maybe this parameter should be mandatory to avoid risk of uploading to the wrong environment by mistake?

**Comment by mwinokan** [comment]
Thanks @kaliif and @tdudgeon. 

The specification of which stack to upload to should always be quite intentional, hence I agree with Tim that the `--url` must always be provided and have no default.
